### PR TITLE
Contributions page

### PR DESCRIPTION
**Pull request by brson**
Based off of [@efindlay's original PR](https://github.com/rust-lang/rust-www/pull/199), which was based off [ncameron's work](http://ncameron.org/rust.html).

[Rendered](http://brson.github.io/rust-www/contribute.html).

The idea here is to quickly provide the most important links (e.g. the second paragraph reroutes people to bug reporting), subdivide the deluge of information in a way that is digestible (six sections - issue triage, docs, community building, tooling, libs, compiler), then guide people to *other sources of information*. I've used prose in an attempt to provide context for newbies to understand the larger picture. Sometimes the same information is provided in different sections, from different perspectives - e.g. fixing easy issues vs. writing easy issues.

Ideally, there is an easy path from here to all relevant information for contributors, but that is not quite true yet, and there are types of information that are just not available. What I *want* this page to do is to link to other, 'living', landing pages for teams, initiatives, topics, etc., and I've tried to do that as much as possible, but there are some categories of information that are not organized anywhere, so sometimes there is more exposition here than I'd like.

This page occasionally links to, and thus endorses, unofficial projects of interest. I want to be clear though that it is not a free-for-all and we are not open to listing arbitrary projects here.

With the quantity and specificity of info here it'll be prone to bitrot and need periodic attendance.

Problems:

* The prominent link to 'an overview of the organization, processes, and policies' links somewhere uncomprehensive. I'd like to write this, but there's a lot of info to cover and I want a break from writing for a bit.
* Topics of blogging and translation aren't covered.
* #rust-libs may not be the right place to direct people who are interested in out-of-tree libs.

I've run out of steam on this and want to get other eyes on it. The libs and tooling sections were the last I hit and it shows.

Although they are very different, my main influence was [Scala's page](http://scala-lang.org/contribute/).

Fixes https://github.com/rust-lang/rust-www/issues/177.

cc @aturon @steveklabnik

### CODE REVIEW

**Comment by brson** [comment]
This links to the [IDEs page](http://rust-lang.org/ides.html), which I have not updated to fit the style of the site. My preference is to move that content to the [dungeon/forge](https://internals.rust-lang.org/t/dungeon-rust-lang-org-secondary-developer-docs/2896).

File: contribute.md [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+
+The day-to-day maintenance of the project revolves around Rust's
+[issue tracker] and [pull requests][PR], and more help is always
+needed. The most basic way to get started contributing to Rust is to
+look for the [E-easy] or [E-mentor] labels. These are meant to be
+approachable for new Rust programmers.
+
+On `E-mentor` issues an experienced Rust developer has volunteered in
+the comments to mentor you through the solving the issue and
+[submitting the fix GitHub pull request][pull]. Contact them about the
```
Comment (badboy) [inline contribute.md]: This sentence needs fixes. I guess at least `the solving the issue` and `submitting the fix GitHub pull request`.

File: contribute.md [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+title: Contributing to Rust &middot; The Rust Programming Language
+---
+
+# Contributing to Rust
+
+You've started learning Rust. You love it, and you want to be a part
+of it. If you're not sure how to get involved, then this page
+will help.
+
+**Just want to report a bug in Rust?** [Follow the bug reporting
```
Comment (steveklabnik) [inline contribute.md]: I would try to rephrase this to lose the "just", the whole point of these pages is that doing this isn't always as obvious as it may seem. Maybe something like "Found a bug and need to report it?"

File: contribute.md [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+The day-to-day maintenance of the project revolves around Rust's
+[issue tracker] and [pull requests][PR], and more help is always
+needed. The most basic way to get started contributing to Rust is to
+look for the [E-easy] or [E-mentor] labels. These are meant to be
+approachable for new Rust programmers.
+
+On `E-mentor` issues an experienced Rust developer has volunteered in
+the comments to mentor you through the solving the issue and
+[submitting the fix GitHub pull request][pull]. Contact them about the
+issue, on the issue tracker by [@mentioning] their name, on IRC, or
```
Comment (steveklabnik) [inline contribute.md]: "their name in a comment"

File: contribute.md [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+<!--
+TODO: weekly triage email?
+TODO: @nrc says suggesting everybody review w/o training is bad
+-->
+
+<a name="docs"></a>
+## Documentation
+
+Documentation is never good enough and there's never enough of it,
+writing documentation is a valuable way to contribute. Many aspects of
```
Comment (steveklabnik) [inline contribute.md]: I would maybe split that first sentence into two, or use a `;` rather than a `,` after "it"

**Comment by brson** [comment]
An idea I liked but didn't implement, mainly because it would add even more stuff, is to put an 'active initiatives' section near the front, that would link to things like the forum posts for RWIB, Glacier, IDEs, MIR, etc. Curious what others think.

File: contribute.md [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+[rfc-issues]: https://github.com/rust-lang/rfcs/issues
+[rfc]: https://github.com/rust-lang/rfcs#table-of-contents
+[rust-lang-nursery]: https://github.com/rust-lang-nursery
+[rust-lang/rust]: https://github.com/rust-lang/rust
+[rust-lang]: https://github.com/rust-lang
+[rust-learning]: https://github.com/ctjhoa/rust-learning
+[rust-rosetta]: https://github.com/Hoverbear/rust-rosetta
+[rustc-guide]: https://github.com/rust-lang/rust/blob/master/src/librustc/README.md
+[rustc-perf]: http://ncameron.org/perf-rustc/
+[rustc-tour]: http://tomlee.co/2014/04/03/a-more-detailed-tour-of-the-rust-compiler/
```
Comment (steveklabnik) [inline contribute.md]: This post is great, but so old: is it even relevant today?

**Comment by steveklabnik** [comment]
:clap: :clap: :clap:

**Comment by nodakai** [comment]
Are we required to read through both of the new contributing.html and the existing CONTRIBUTING.md ?  I don't think we need both.

File: contribute.md [hunk 6]
```diff
@@ -0,0 +1,454 @@
+---
+layout: default
+title: Contributing to Rust &middot; The Rust Programming Language
+---
+
+# Contributing to Rust
```
Comment (nodakai) [inline contribute.md]: "Contributing to Rust" carries next to no information.  Instead I'd emphasize

> There are many ways to contribute to the success of Rust.

because, as I understand it, this document is intended for people who are eager to contribute to Rust but have no idea on where to start.

**Comment by englishm** [comment]
How might people with UX or visual design backgrounds be able to contribute to Rust?

File: contribute.md [hunk 7]
```diff
@@ -0,0 +1,454 @@
...
+Rust is always in need of people to [triage] issues: reproduce bugs,
+minimize test cases, apply labels, close resolved issues. Note that
+you'll need elevated GitHub permissions to apply labels, but this is
+easy to obtain for somebody with a bit of experience in the
+project. Ask a [team member][team].
+
+Once you've found your way around the project and have created a few
+pull requests in a particular area, consider reviewing others' pull
+requests: good reviewership is a rare skill and always appreciated. No
+prior permission is needed &emdash; just start constructively and politely
```
Comment (iliekturtles) [inline contribute.md]: The `&emdash;` is not rendering properly: search for "emdash" at http://brson.github.io/rust-www/contribute.html

File: contribute.md (hunk 1 above)
Comment (brson) [inline contribute.md]: Wow, what happened...

File: contribute.md (hunk 4 above)
Comment (brson) [inline contribute.md]: The second clause is self-evident and content free, so I think I'll just remove it.

File: contribute.md (hunk 5 above)
Comment (brson) [inline contribute.md]: Maybe not. I'll remove it.

**Comment by brson** [comment]
> Are we required to read through both of the new contributing.html and the existing CONTRIBUTING.md ? I don't think we need both.

CONTRIBUTING.md is displayed on new PRs - it catches people who are familiar with the GitHub ecosystem but not the Rust project. It is also specific to the *mechanics* of contributing to the rust-lang/rust repository. This page is much broader in scope and links to CONTRIBUTING.md as appropriate for the details.

The content is mostly not duplicated, but I think we can trim down CONTRIBUTING.md.

File: contribute.md (hunk 6 above)
Comment (brson) [inline contribute.md]: What might be a more descriptive title?

**Comment by brson** [comment]
> How might people with UX or visual design backgrounds be able to contribute to Rust?

@englishm Unfortunately, we don't have an answer for this right now. We're hoping to put some planning effort into this area soon so we can give better guidance.

File: contribute.md (hunk 7 above)
Comment (brson) [inline contribute.md]: Oops. Should be mdash.

**Comment by brson** [comment]
I've addressed the current feedback except for @nodakai's concern about the title.

**Comment by aturon** [comment]
I want to give more in-depth feedback, but for now, my biggest concern is that there's just *too* much here -- that the page is a bit overwhelming.

Can we think of a way to pull out the most juicy bits into a TL;DR section we can stick at the front? I know triage is acting as part of that, but I'm thinking something even tighter.

Note that the Scala site avoids this problem by having all of the sections actually be independent pages. We could consider that as well.

**Comment by brson** [comment]
@aturon Yeah, it's a lot. Do you think that the six areas are a reasonable subdivision?

@bstrie has previously expressed an opinion against subdividing the pages.

I'll try reorganizing it so this page is less sprawling, with the bulk of the information elsewhere.

**Comment by brson** [comment]
cc @nrc you may have opinions about this.

**Comment by aturon** [comment]
@brson Yes, I think the structure and content is great. Just hoping we can get something a bit more inviting *a la* http://scala-lang.org/contribute/

**Comment by nrc** [comment]
+1 for subpages - I think it will help make the page feel less overwhelming. Design-wise, I think having more whitespace and a less dense layout will make this feel better, and that is easier with sub-pages, if the landing page has just a few paras of text and some links, it's easier to space it out.

**Comment by brson** [comment]
OK, I've split each of the 6 subsections into their own pages. [See here](http://brson.github.io/rust-www/contribute.html).

**Comment by aturon** [comment]
@brson **Very nice**! I feel like that really makes this whole effort come together.

**Comment by aturon** [comment]
One further suggestion: make the links to the subsections stand out a bit -- perhaps bold? There are a lot of links here so emphasizing the primary ones is helpful. (See, again, the Scala page for inspiration here.)

**Comment by brson** [comment]
I've emboldened the subsection links and also removed the prominent governance link temporarily until we have content for it.

**Comment by aturon** [comment]
@brson: :+1:

Btw, are you planning to land these various PRs separately? Or, now that there are cross-cutting style changes, do you want to land them all together?

**Comment by brson** [comment]
@aturon At this point, with the style changes that affect everything I'm planning on landing them together.

**Comment by brson** [comment]
[Combined PR](https://github.com/rust-lang/rust-www/pull/244).
